src/popup/components/Modals/QrCodeReader.vue

```diff
@@ -5,201 +5,270 @@
-import BtnMain from '../buttons/BtnMain.vue';
-import QrScanIcon from '../../../icons/qr-scan.svg?vue-component';
-import IconBoxed from '../IconBoxed.vue';
+import AnimatedSpinnerIcon from '@/icons/animated-spinner.svg?skip-optimize';
+import QrScanIcon from '@/icons/qr-scan.svg?vue-component';
 
-export default {
+export default defineComponent({
   components: {
     Modal,
     BtnMain,
     IconBoxed,
+    AnimatedSpinnerIcon,
   },
   props: {
     title: { type: String, required: true },
     resolve: { type: Function, required: true },
     reject: { type: Function, required: true },
   },
-  setup() {
-    return {
-      QrScanIcon,
-    };
-  },
-  data: () => ({
-    mobile: IS_CORDOVA,
+  setup(props) {
     // allow camera while QRScanner is loading to not show cameraNotAllowed before actual check
-    cameraAllowed: IS_CORDOVA,
-    browserReader: null,
-    videoInputDevices: [],
-    headerText: '',
-  }),
-  watch: {
-    async cameraAllowed(value) {
-      if (!value) {
-        this.stopReading();
+    const cameraStatus = ref<PermissionState>(
+      IS_MOBILE_APP ? 'granted' : 'denied',
+    );
+    const browserReader = ref<BrowserQRCodeReaderType | null>(null);
+    const browserReaderControls = ref<IScannerControls>();
+    const qrCodeVideoEl = ref<HTMLVideoElement>();
+
+    const route = useRoute();
+    const { t } = useI18n();
+    const { setQrScanner } = useUi();
+
+    const cameraAllowed = computed(() => cameraStatus.value === 'granted');
+    const heading = computed(() => {
+      if (cameraStatus.value === 'granted') return t('modals.qrCodeReader.scanQr');
+      return t('modals.qrCodeReader.grantPermission');
+    });
+
+    async function initBrowserReader() {
+      const { BrowserQRCodeReader } = await import('@zxing/browser');
+      browserReader.value = new BrowserQRCodeReader();
+    }
+
+    function stopReading() {
+      if (IS_MOBILE_APP) {
+        document.querySelector('body')?.classList.remove('scanner-active');
+        BarcodeScanner.showBackground();
+        setQrScanner(false);
+        BarcodeScanner.stopScan();
+      } else {
+        browserReaderControls.value?.stop();
+      }
+    }
+
+    async function scan() {
+      if (IS_MOBILE_APP) {
+        return new Promise((resolve, reject) => {
+          setQrScanner(true);
+
+          document.querySelector('body')?.classList.add('scanner-active');
+          BarcodeScanner.hideBackground().then(() => {
+            BarcodeScanner.startScan().then((result) => {
+              if (result.hasContent) {
+                resolve(result.content);
+              } else {
+                reject(new Error('No content'));
+              }
+            }, (error) => {
+              reject(error);
+            });
+          });
+
+          setTimeout(() => {
+            document.querySelector('.camera-close-button')?.addEventListener('click', stopReading);
+          }, 500);
+        });
+      }
+      return new Promise((resolve) => {
```

Comment:
Here we actually want the `Promise` because `decodeFromVideoDevice` doesn't return the result in a promise but rather we have to use a callback method